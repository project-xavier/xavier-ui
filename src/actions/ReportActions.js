import { getAllReports, getReportById } from '../api/report';

export const FETCH_REPORTS = 'FETCH_REPORTS';
export const FETCH_REPORT = 'FETCH_REPORT';

export const fetchReports = () => ({
    type: FETCH_REPORTS,
    payload: getAllReports(),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: 'Failed to load reports'
            }
        }
    }
});

export const fetchReport = (id) => ({
    type: FETCH_REPORT,
    payload: getReportById(id),
    meta: {
        notifications: {
            rejected: {
                variant: 'danger',
                title: `Failed to load report ${ id }`
            }
        }
    }
});
