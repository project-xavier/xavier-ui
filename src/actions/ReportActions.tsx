import {
  getAllReports,
  getReportById,
  deleteReport as destroyReport,
  getReportWokloadSummary,
  getReportWorkloadsDetected,
  getReportFlags,
  getReportInitialSavingestimation,
  getReportWorkloadInventory,
  getReportWorkloadInventoryAllCSV,
  getReportWorkloadInventoryFilteredCSV,
  getReportWorkloadInventoryAvailableFilters,
  getReportPayloadDownloadLink,
} from '../api/report';
import { GenericAction } from '../models/action';

export const ActionTypes = {
  FETCH_REPORTS: 'FETCH_REPORTS',
  FETCH_REPORT: 'FETCH_REPORT',
  DELETE_REPORT: 'DELETE_REPORT',
  FETCH_REPORT_WORKLOAD_SUMMARY: 'FETCH_REPORT_WORKLOAD_SUMMARY',
  FETCH_REPORT_WORKLOADS_DETECTED: 'FETCH_REPORT_WORKLOADS_DETECTED',
  FETCH_REPORT_FLAGS: 'FETCH_REPORT_FLAGS',
  FETCH_REPORT_INITIAL_SAVING_ESTIMATION: 'FETCH_REPORT_INITIAL_SAVING_ESTIMATION',
  FETCH_REPORT_WOKLOAD_INVENTORY: 'FETCH_REPORT_WOKLOAD_INVENTORY',
  FETCH_REPORT_WOKLOAD_INVENTORY_ALL_CSV: 'FETCH_REPORT_WOKLOAD_INVENTORY_ALL_CSV',
  FETCH_REPORT_WOKLOAD_INVENTORY_FILTERED_CSV: 'FETCH_REPORT_WOKLOAD_INVENTORY_FILTERED_CSV',
  FETCH_REPORT_WOKLOAD_INVENTORY_AVAILABLE_FILTERS:
    'FETCH_REPORT_WOKLOAD_INVENTORY_AVAILABLE_FILTERS',
  FETCH_REPORT_PAYLOAD_DOWNLOAD_LINK: 'FETCH_REPORT_PAYLOAD_DOWNLOAD_LINK',
};

/**
 * Fetchs reports
 */
export const fetchReports = (page: number, perPage: number, filterText: string): GenericAction => {
  return {
    type: ActionTypes.FETCH_REPORTS,
    payload: getAllReports(page, perPage, filterText),
    meta: {
      notifications: {
        rejected: {
          variant: 'danger',
          title: 'Failed to load reports',
        },
      },
    },
  };
};

/**
 * Fetchs a single report
 * @param id report ID
 */
export const fetchReport = (id: number): GenericAction => ({
  type: ActionTypes.FETCH_REPORT,
  payload: getReportById(id),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to load report ${id}`,
      },
    },
  },
});

export const deleteReport = (id: number, name: string): GenericAction => ({
  type: ActionTypes.DELETE_REPORT,
  payload: destroyReport(id).then(() => ({ id })),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to delete report ${name}`,
      },
      fulfilled: {
        variant: 'success',
        title: `Report ${name} deleted`,
      },
    },
  },
});

export const fetchReportWorkloadSummary = (id: number): GenericAction => ({
  type: ActionTypes.FETCH_REPORT_WORKLOAD_SUMMARY,
  payload: getReportWokloadSummary(id),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to load report workload summary ${id}`,
      },
    },
  },
});

export const fetchReportWorkloadsDetected = (
  id: number,
  page: number,
  perPage: number,
  orderBy: string | undefined,
  orderDirection: 'asc' | 'desc' | undefined
): GenericAction => ({
  type: ActionTypes.FETCH_REPORT_WORKLOADS_DETECTED,
  payload: getReportWorkloadsDetected(id, page, perPage, orderBy, orderDirection),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to load report workloads detected ${id}`,
      },
    },
  },
});

export const fetchReportFlags = (
  id: number,
  page: number,
  perPage: number,
  orderBy: string | undefined,
  orderDirection: 'asc' | 'desc' | undefined
): GenericAction => ({
  type: ActionTypes.FETCH_REPORT_FLAGS,
  payload: getReportFlags(id, page, perPage, orderBy, orderDirection),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to load report flags ${id}`,
      },
    },
  },
});

export const fetchReportInitialSavingEstimation = (id: number): GenericAction => ({
  type: ActionTypes.FETCH_REPORT_INITIAL_SAVING_ESTIMATION,
  payload: getReportInitialSavingestimation(id),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to load report saving estimation ${id}`,
      },
    },
  },
});

export const fetchReportWorkloadInventory = (
  id: number,
  page: number,
  perPage: number,
  orderBy: string | undefined,
  orderDirection: 'asc' | 'desc' | undefined,
  filters: Map<string, string[]>
): GenericAction => ({
  type: ActionTypes.FETCH_REPORT_WOKLOAD_INVENTORY,
  payload: getReportWorkloadInventory(id, page, perPage, orderBy, orderDirection, filters),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to load report workload inventory ${id}`,
      },
    },
  },
});

export const fetchReportWorkloadInventoryAllCSV = (id: number): GenericAction => ({
  type: ActionTypes.FETCH_REPORT_WOKLOAD_INVENTORY_ALL_CSV,
  payload: getReportWorkloadInventoryAllCSV(id),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to load report workload inventory ${id} csv`,
      },
    },
  },
});

export const fetchReportWorkloadInventoryFilteredCSV = (
  id: number,
  orderBy: string | undefined,
  orderDirection: 'asc' | 'desc' | undefined,
  filters: Map<string, string[]>
): GenericAction => ({
  type: ActionTypes.FETCH_REPORT_WOKLOAD_INVENTORY_FILTERED_CSV,
  payload: getReportWorkloadInventoryFilteredCSV(id, orderBy, orderDirection, filters),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to load report workload inventory ${id} csv`,
      },
    },
  },
});

export const fetchReportWorkloadInventoryAvailableFilters = (id: number): GenericAction => ({
  type: ActionTypes.FETCH_REPORT_WOKLOAD_INVENTORY_AVAILABLE_FILTERS,
  payload: getReportWorkloadInventoryAvailableFilters(id),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to load report workload inventory filters ${id}`,
      },
    },
  },
});

export const fetchReportPayloadDownloadLink = (id: number): GenericAction => ({
  type: ActionTypes.FETCH_REPORT_PAYLOAD_DOWNLOAD_LINK,
  payload: getReportPayloadDownloadLink(id),
  meta: {
    notifications: {
      rejected: {
        variant: 'danger',
        title: `Failed to fetch report ${id} payload download link`,
      },
    },
  },
});
