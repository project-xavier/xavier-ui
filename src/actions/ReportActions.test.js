jest.mock('../api/apiClient', () => ({
    get: jest.fn(() => Promise.resolve({})),
    post: jest.fn(() => Promise.resolve({})),
    put: jest.fn(() => Promise.resolve({})),
    delete: jest.fn(() => Promise.resolve({}))
}));

import * as actionGenerator from './ReportActions';

describe('fetchReports', () => {
    it('returns a state object', () => {
        expect(actionGenerator.fetchReports()).toMatchSnapshot();
    });
});

describe('fetchReport', () => {
    it('returns a state object', () => {
        expect(actionGenerator.fetchReport(1)).toMatchSnapshot();
    });
});
