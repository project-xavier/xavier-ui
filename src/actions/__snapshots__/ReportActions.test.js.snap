// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchReport returns a state object 1`] = `
Object {
  "meta": Object {
    "notifications": Object {
      "rejected": Object {
        "title": "Failed to load report 1",
        "variant": "danger",
      },
    },
  },
  "payload": Promise {},
  "type": "FETCH_REPORT",
}
`;

exports[`fetchReports returns a state object 1`] = `
Object {
  "meta": Object {
    "notifications": Object {
      "rejected": Object {
        "title": "Failed to load reports",
        "variant": "danger",
      },
    },
  },
  "payload": Promise {},
  "type": "FETCH_REPORTS",
}
`;
